docs/enhancements/cluster-pool-inventory.md

```diff
@@ -0,0 +1,158 @@
+    - [How To Use](#how-to-use)
+    - [Validation](#validation)
+    - [`Size` and `MaxSize`](#size-and-maxsize)
+    - [Pool Version](#pool-version)
+    - [Handling Inventory Updates](#handling-inventory-updates)
+      - [Adding An Inventory](#adding-an-inventory)
+      - [Adding An Entry](#adding-an-entry)
+      - [Removing An Entry](#removing-an-entry)
+      - [Deleting The Inventory](#deleting-the-inventory)
+    - [Challenge: Namespace Collision](#challenge-namespace-collision)
+
+## User Stories
+As a cluster administrator, I want to be able to use ClusterPools to provision VSphere clusters.
+
+(Future) As a cluster administrator, I want to be able to use ClusterPools to provision bare metal clusters.
+
+## Problem Statement
+Provisioning a cluster requires configuration unique to that cluster.
+For example, the cluster's name is used to build the hostnames for the API and console.
+_Someone_ has to handle the DNS that resolves those hostnames to the IPs of the services created by the installer.
+One option is to create your ClusterDeployment with `ManageDNS=True` and we'll create a DNSZone on the fly.
+
+But that doesn't work for some cloud providers, such as VSphere, the focus of this document.
+
+In the case of VSphere, one solution is to configure DNS manually and then create your ClusterDeployment with the right `Name` and `BaseDomain` so that the assembled hostnames match those entries. This is fine if you're creating ClusterDeployments by hand, but breaks down if ClusterDeployment names are being generated with random slugs, as is the case with ClusterPools.
+
+## Scope
+This feature is limited to solving DNS for VSphere.
+Bare metal will come later, _but_ we will try to be forward-thinking in this design to allow for future expansion to cover that user story.
+
+## Proposal
+### Summary
+Allow ClusterPool to accept an inventory of per-cluster details to be used when generating ClusterDeployments for the pool.
+
+### `ClusterPool.Spec.Inventory`
+Add a field to `ClusterDeployment.Spec` called `Inventory`.
+It is a dict.
+For now, the dict can have one member, `ClusterDeployments`, which is a list of dicts.
+For now, each dict in the list can have one member, `Name`, which is a string.
+Example:
+
+```yaml
+spec:
+  inventory:
+    clusterDeployments:
+    - name: foo
+    - name: bar
+    - name: baz
+```
+
+(This may seem overengineered, but the idea is to leave room for future expansion.
+For example, perhaps we'll need to specify IP addresses that are tied to each name.
+If we designed the inventory as simply a list of name strings, there would be nowhere to put those IPs.)
+
+When adding a ClusterDeployment, if such an `Inventory` is present, ClusterPool will:
+- Load up the inventory list.
+- Load up the existing pool CDs (it already does this).
+- Subtract the set of names of existing CDs from those in the inventory list.
+- Pick a name from the set that remains and use it for the new CD.
```

Comment:
This code will be in a controller that watches ClusterPool, giving you some guarantee that we're not concurrently doing this calculation for more than one request and might select the same name twice?

Probably will want to watch in the calculation for two hits for the same name and do... something... I don't know what. :)

Reply:
This kind of implementation kind of guarantees that we can only run one replica of hive-controllers, right?

Reply:
Leader election ensures that only one of those pods is actively processing at a time even if you're using multiple replicas. They're more like hot backups which is sort of why we weren't super worried about the multiple replicas.

That said Thomas' clustersync thing we did for multiple pods each acting on only a subset if the hash matches is another option we might use in future, but even there it should be close to impossible for two pods to reconcile one object in the same controller. We don't have full guarantees here as we're not a transactional db, but it should be pretty reliable day to day, you just might want to keep an eye out for it going wrong and react or warn somehow.

Reply:
Worst case we could explicitly keep track -- e.g. ClusterPool.Status.ClusterNamesInUse, and bouncing the status update with a 409 is our concurrency protection -- but I'd like to avoid that if possible. If we can solve the namespace collision thing in a way that lets us keep the names deterministic, that would also work.

Still thinking this through...

Reply:
After a week of noodling, I think I'd like to address this with an annotation containing a JSON dict, keyed by the cluster name from the inventory list, of the namespace of the CD consuming it.

- When the value is empty/null/nil, that name hasn't been claimed yet.
- When we need to create a new cluster, we
  - Scan the list for an entry with a nil value
  - Generate a unique namespace name using that name as a stub
  - Post an update to the CD with the annotation updated with that namespace as the value for that entry. Here's where we bounce if two controllers try to grab the same name at the same time: the loser will bounce the update with a 409; we'll requeue and start over.
  - Proceed with existing algorithm to create new cluster (except setting cluster name explicitly as noted [below](https://github.com/openshift/hive/pull/1549#discussion_r719721440)).
- Add a top-level routine to ensure all replete entries have CDs associated with them. If not, attempt to create them through the same flow. This is to guard against leaking entries if a controller crashes between updating the annotation and creating the CD.

**Question:** Sorting. Is it important that we rotate through the list of supplied cluster names in any particular order, or with any nod to "fairness"? For example: If there are five names and the usage pattern happens to only use two at a time, is it okay to reuse the same two over and over, or must we strive to consume all of them evenly?

Reply:
Per [slack discussion](https://coreos.slack.com/archives/CE3ETN3J8/p1634157465296300), moved to a status field.